cmd/influxd/upgrade/upgrade.go

```diff
@@ -44,76 +83,476 @@ var options = struct {
+			return err
+		}
+		v1Config, err := upgradeConfig(options.source.configFile, log)
+		if err != nil {
+			return err
+		}
+		// TODO how to load and validate new Config???
+
+		options.source.metaDir = v1Config.Meta.Dir
+		options.source.dataDir = v1Config.Data.Dir
+		options.source.walDir = v1Config.Data.WALDir
+	} else {
+		log.Info("no InfluxDB 1.x config file specified, skipping its upgrade")
+	}
+
+	if err := options.source.checkDirs(); err != nil {
+		return err
+	}
 
 	v1, err := newInfluxDBv1(&options.source)
 	if err != nil {
 		return err
 	}
-	_ = v1
 
-	v2, err := newInfluxDBv2(ctx, &options.target)
+	v2, err := newInfluxDBv2(ctx, &options.target, log)
 	if err != nil {
 		return err
 	}
-	_ = v2
 
-	// 1. Onboard the initial admin user
-	// v2.onboardSvc.OnboardInitialUser()
+	or, err := setupAdmin(ctx, v2)
+	if err != nil {
+		return err
+	}
+	options.target.token = or.Auth.Token
 
-	// 2. read each database / retention policy from v1.meta and create bucket db-name/rp-name
-	// newBucket := v2.ts.CreateBucket(ctx, Bucket{})
-	//
-	// 3. create database in v2.meta
-	// v2.meta.CreateDatabase(newBucket.ID.String())
-	// copy shard info from v1.meta
+	db2BucketIds, err := upgradeDatabases(ctx, v1, v2, or.Org.ID, log)
+	if err != nil {
+		//remove all files
+		log.Info("database upgrade error, removing data")
+		if e := os.Remove(options.target.boltPath); e != nil {
+			log.Error("cleaning failed", zap.Error(e))
+		}
+
+		if e := os.RemoveAll(options.target.enginePath); e != nil {
+			log.Error("cleaning failed", zap.Error(e))
+		}
+		return err
+	}
+
+	if err = generateSecurityScript(v1, db2BucketIds, log); err != nil {
+		return err
+	}
+
+	log.Info("upgrade successfully completed. Start service now")
 
 	return nil
 }
 
+func upgradeDatabases(ctx context.Context, v1 *influxDBv1, v2 *influxDBv2, orgID influxdb.ID, log *zap.Logger) (map[string][]string, error) {
+	db2BucketIds := make(map[string][]string)
+	targetDataPath := filepath.Join(options.target.enginePath, "data")
+	targetWalPath := filepath.Join(options.target.enginePath, "wal")
+	dirFilterFunc := func(path string) bool {
+		base := filepath.Base(path)
+		if base == "_series" ||
+			(len(base) > 0 && base[0] == '_') || //skip internal databases
+			base == "index" {
+			return true
+		}
+		return false
+	}
+	// read each database / retention policy from v1.meta and create bucket db-name/rp-name
+	// create database in v2.meta
+	// copy shard info from v1.meta
+	if len(v1.meta.Databases()) > 0 {
+		// Check space
+		log.Info("checking space")
+		size, err := fs2.DirSize(options.source.dataDir)
+		if err != nil {
+			return nil, fmt.Errorf("error getting size of %s: %w", options.source.dataDir, err)
+		}
+		size2, err := fs2.DirSize(options.source.walDir)
+		if err != nil {
+			return nil, fmt.Errorf("error getting size of %s: %w", options.source.walDir, err)
+		}
+		size += size2
+		v2dir := filepath.Dir(options.target.boltPath)
+		diskInfo, err := fs2.DiskUsage(v2dir)
```

Comment:
Rather than import this very OS-dependent code in order to do a fairly inaccurate check (sum of file sizes doesn't equate to used disk space), it might be best just to let the user decide for themselves and provide a "dry run" mode that just says how much new disk space will be needed or something.

Reply:
I suppose user databases  have size in hundreads of MB and probably more.. Do you think that it will be such a big difference that these will give false result?

Reply:
Actual error can be just because filesystem entries size and block size rounding, but comparing to the datafile sizes this shouldn't be much difference. 
Even available disk space can be different during copying data, due to the other processes work. 
This just a sanity check. If disk available size is very close to the requires size (condition when the check inaccuracy matters) is even bad for a system.

On the other hand, _dry run_ mode is a new feature and can be added later..